crates/sui-network/src/state_sync.rs

```diff
@@ -0,0 +1,1433 @@
+                self.store.clone(),
+                self.peer_heights.clone(),
+                self.checkpoint_event_sender.clone(),
+                // The if condition should ensure that this is Some
+                highest_known_checkpoint.unwrap(),
+            )
+            .map(|maybe_checkpoint| match maybe_checkpoint {
+                Ok(checkpoint) => Some(checkpoint),
+                Err(e) => {
+                    debug!("error syncing checkpoint {e}");
+                    None
+                }
+            });
+            self.sync_tasks.spawn(task);
+        }
+    }
+
+    fn gossip_checkpoint(&mut self, checkpoint: VerifiedCheckpoint) {
+        let task = gossip_checkpoint(self.network.clone(), self.peer_heights.clone(), checkpoint);
+        self.tasks.spawn(task);
+    }
+}
+
+async fn gossip_checkpoint(
+    network: anemo::Network,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    checkpoint: VerifiedCheckpoint,
+) {
+    let futs = peer_heights
+        .read()
+        .unwrap()
+        .heights
+        .iter()
+        // Filter out any peers who we know already have a checkpoint higher than this one
+        .filter(|(_peer_id, &height)| Some(checkpoint.sequence_number()) > height)
+        .map(|(peer_id, _height)| peer_id)
+        .flat_map(|peer_id| network.peer(*peer_id))
+        .map(StateSyncClient::new)
+        .map(|mut client| {
+            let request = Request::new(checkpoint.inner().clone()).with_timeout(DEFAULT_TIMEOUT);
+            async move { client.push_checkpoint_summary(request).await }
+        })
+        .collect::<Vec<_>>();
+    futures::future::join_all(futs).await;
+}
+
+async fn get_latest_from_peer(
+    peer: anemo::Peer,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    tx: mpsc::WeakSender<StateSyncMessage>,
+) {
+    let peer_id = peer.peer_id();
+    let request = Request::new(GetCheckpointSummaryRequest::Latest).with_timeout(DEFAULT_TIMEOUT);
+    let response = StateSyncClient::new(peer)
+        .get_checkpoint_summary(request)
+        .await
+        .map(Response::into_inner);
+    update_peer_height(&peer_heights, peer_id, &response);
+
+    //TODO maybe only send event if its newer?
+    if let Ok(Some(_checkpoint)) = response {
+        if let Some(tx) = tx.upgrade() {
+            let _ = tx.send(StateSyncMessage::StartSyncJob).await;
+        }
+    }
+}
+
+fn update_peer_height(
+    peer_heights: &RwLock<PeerHeights>,
+    peer_id: PeerId,
+    response: &Result<Option<Checkpoint>, Status>,
+) {
+    match response {
+        Ok(latest) => {
+            peer_heights
+                .write()
+                .unwrap()
+                .update_peer_height(peer_id, latest.clone());
+        }
+        Err(status) => {
+            // TODO maybe don't clear on timeout?
+            trace!("get_latest_checkpoint_summary request failed: {status:?}");
+            peer_heights.write().unwrap().heights.remove(&peer_id);
+        }
+    }
+}
+
+async fn handle_tick(
```

Comment:
my first thought: tick of what? for a top level function this name is kinda not helpful enough (or maybe just a case of needs more comments/more smaller files again)